core/computed/entity-classification.js

```diff
@@ -0,0 +1,114 @@
+   * @return {LH.Artifacts.Entity | undefined}
+   */
+  static makeUpAnEntity(entityCache, url) {
+    if (!UrlUtils.isValid(url)) return;
+    // We can make up an entity only for those URLs with a valid domain attached.
+    // So we further restrict from allowed URLs to (http/https).
+    if (!Util.createOrReturnURL(url).protocol.startsWith('http')) return;
+
+    const rootDomain = Util.getRootDomain(url);
+    if (!rootDomain) return;
+    if (entityCache.has(rootDomain)) return entityCache.get(rootDomain);
+
+    const unrecognizedEntity = {
+      name: rootDomain,
+      company: rootDomain,
+      category: '',
+      categories: [],
+      domains: [rootDomain],
+      averageExecutionTime: 0,
+      totalExecutionTime: 0,
+      totalOccurrences: 0,
+      isUnrecognized: true,
+    };
+    entityCache.set(rootDomain, unrecognizedEntity);
+    return unrecognizedEntity;
+  }
+
+  /**
+   * @param {{URL: LH.Artifacts['URL'], devtoolsLog: LH.DevtoolsLog}} data
+   * @param {LH.Artifacts.ComputedContext} context
+   * @return {Promise<LH.Artifacts.EntityClassification>}
+   */
+  static async compute_(data, context) {
+    const networkRecords = await NetworkRecords.request(data.devtoolsLog, context);
+    /** @type {EntityCache} */
+    const madeUpEntityCache = new Map();
+    /** @type {Map<string, LH.Artifacts.Entity>} */
+    const entityByUrl = new Map();
+    /** @type {Map<LH.Artifacts.Entity, Array<string>>} */
+    const urlsByEntity = new Map();
+
+    for (const record of networkRecords) {
+      const {url} = record;
+      if (entityByUrl.has(url)) continue;
+
+      const entity = thirdPartyWeb.getEntity(url) ||
+        EntityClassification.makeUpAnEntity(madeUpEntityCache, url);
+      if (!entity) continue;
+
+      const entityURLs = urlsByEntity.get(entity) || [];
+      entityURLs.push(url);
+      urlsByEntity.set(entity, entityURLs);
+      entityByUrl.set(url, entity);
+    }
+
+    // When available, first party identification will be done via
+    // `mainDocumentUrl` (for navigations), and falls back to `finalDisplayedUrl` (for timespan/snapshot).
+    // See https://github.com/GoogleChrome/lighthouse/issues/13706
+    /** @type {LH.Artifacts.Entity | undefined} */
+    let firstParty;
+    const firstPartyUrl = data.URL.mainDocumentUrl || data.URL.finalDisplayedUrl;
+    if (firstPartyUrl) {
+      firstParty = thirdPartyWeb.getEntity(firstPartyUrl) ||
+        EntityClassification.makeUpAnEntity(madeUpEntityCache, firstPartyUrl);
+    }
+
+    /**
+     * Convenience function to return the entity name for a URL.
+     * @param {string} url
+     * @return {string | undefined}
+     */
+    function getEntityName(url) {
+      return entityByUrl.get(url)?.name;
+    }
+
+    /**
+     * Convenience function to check if a URL belongs to first party.
+     * @param {string} url
+     * @return {boolean}
+     */
+    function isFirstParty(url) {
+      return entityByUrl.get(url) === firstParty;
```

Comment:
It's possible for `firstParty` and `entityByUrl.get(url)` to both be undefined right? Do we want to return true in that case?

Reply:
That's an interesting scenario. `entityByUrl` not having a URL should be a bug, do you foresee a scenario?

Reply:
Wait `firstParty` should always be defined because we use `makeUpAnEntity`

Reply:
Yeah if `finalDisplayedUrl` is always present, and it is going to be a valid `http[s]` url, then it should be defined.

Reply:
But if it's not a valid `http[s]` then `firstParty` will be undefined? Maybe it's a good idea to have a defensive `if(!firstParty) return false` in this function.

Reply:
> `entityByUrl` not having a URL should be a bug, do you foresee a scenario?

it definitely feels like there should be a fallback or a loud error for this case, e.g. if I made this request, would it have been first/third party. I don't think there's anywhere where we do this, but I think it's very possible someone wouldn't notice if they added it in the future and this just always returns `false` for that case.

A `if (!entityByUrl.get(url)) throw new Error('...')` equivalent would probably be fine to keep things simple until it's actually needed.

Reply:
I added a `throw` in case we have urls being presented, that are not part of `devtoolsLog`. This was a concern we had before anyway, so makes sense to fail loud in that case. If first-party entity is unavailable, the equality would fail and we'd return `false` anyway so should work for that as well.